### PR TITLE
Refactor to add an interface for view_state of #mrheader

### PR DESCRIPTION
**Pull request by eiri**
The goal of this refactoring is to make it easier to understand and maintain `view_state` data structure of `#mrheader`.

Unlike record `#mrview` in `#mrst`, the data structure that represents `view_states` in `#mrheader` is a tuple (of tuples) and its definition and application spreads across all `couch_mrview_util.erl` module.
This makes it hard to follow and understand how it used or troubleshoot mrview in case of the problems.

This change introduces an interface for generation and accessing `view_state` and also separates generation of reduction closures from mrview's `reduce`, `reduce_to_count` and `fold_reduce` functions as they rely on view_state's structure.

### CODE REVIEW

**Comment by eiri** [comment]
@davisp: hate to bother you, but you know mrview the best.
Can you please take a look on this if you have a bit of free time and let me know what you think?

I was dealing with mrview lately and `view_state` was one of the most subtle (like in "hard to grasp") and error prune parts of it, so I want to do something about it.

**Comment by kxepal** [comment]
@eiri Nice work! But is it possible to get rid anonymous functions? I see it's possible at least for `make_counts_reduce_fun` case, may be for else too.

**Comment by eiri** [comment]
@kxepal Thank you!

I'm afraid it's not feasible to completely get rid of closures there. We are calling functor `couch_btree:final_reduce/2` which accepts a function of arity 2 and invokes it with tag and either list of key/values or result of previous reductions, no other parameters. Since we also need to know our view's reduction function source and language to call  `couch_query_servers:reduce/2` inside of that passed functions, we have to wrap them in a closure.
One alternative would be to call index server to get function source from its state, but that'd quickly get expensive from performance perspective.

Good catch on `make_counts_reduce_fun/0` though, I've got carried away by refactoring consistency there, making all into generators :) I'll change it into function of the module instead of a generator.

File: src/couch_mrview_util.erl [hunk 1]
```diff
@@ -341,18 +323,9 @@ all_docs_reduce_to_count(Reductions) ->
...
-            ],
-            {lists:sum(Counts), []};
-        (rereduce, Reds) ->
-            {lists:sum([Count0 || {Count0, _} <- Reds]), []}
-    end,
-    {Count, _} = couch_btree:final_reduce(Reduce, Reductions),
-    Count.
+    CountsReduceFun = fun counts_reduce/2,
+    FinalReduction = couch_btree:final_reduce(CountsReduceFun, Reductions),
+    get_counts(FinalReduction).
```
Comment (kxepal) [inline src/couch_mrview_util.erl]: A bit naming issue: `reduce_to_count` means that we expect singular count e.g. a number. Before refactoring there was also `Count` word around. But here we use `counts` which implicitly assumes many numbers e.g. list of numbers which is actually not. Could be confusing.

**Comment by kxepal** [comment]
@eiri good, thanks for explanation. LGFM except one minor naming issue.

**Comment by iilyak** [comment]
@eiri: Awesome work. Thank you for doing this.

**Comment by iilyak** [comment]
Tested locally, works as expected.

**Comment by eiri** [comment]
@kxepal, @iilyak - Thank you for the review!

I'll normalize count naming before of the squash.

**Comment by kxepal** [comment]
+1

**Comment by eiri** [comment]
Merged as per [6ec35971ebd96133cc6a34fa29f1d364049a7ecd](https://git-wip-us.apache.org/repos/asf?p=couchdb-couch-mrview.git;a=commit;h=6ec35971ebd96133cc6a34fa29f1d364049a7ecd).

**Comment by kxepal** [comment]
@eiri Hint: rebase, push to your PR branch, push to master -> PR get closed automagically as merged (:

Thanks anyway! \o/

**Comment by eiri** [comment]
@kxepal - either squash or changing of first PR's description (like fixing of typos ;) apparently confusing that hook, I've seen that the merge got synced to github properly, but PR was left open.
